### PR TITLE
Generalize submitter, .env, and bash and slurm scripts

### PR DESCRIPTION
**Pull request by egrace479**
The sample `.env` file now defines paths previously defined in the submitted script. It also has the project name, so that and all the distribution of work settings can be passed to the `slurm` scripts directly through the bash coordination script (`example_scripts/submit_mpi_download.sh`) instead of needing to be redefined at the top of each file. This does add a slightly higher resource request for the verifier, but (following our group discussion) this shouldn't add much to resource allocation since it is only run for a couple seconds on each batch.

`src/submitter.py` is now the primary user interface with the package (following pre-processing steps). Its only input is the path to the `.env` file, so that it can fetch the required paths. There was also some restructuring to avoid redefining variables.

### CODE REVIEW

File: src/submitter.py [hunk 1]
```diff
@@ -10,13 +12,34 @@
 RECHECK = False
```
Comment (thompsonmj) [inline src/submitter.py:12]: It won't let me comment on the line above for some reason, but `NUM_DOWNLOADERS` should also go into `hpc.env` to be configurable.

File: example_scripts/submit_mpi_download.sh
Comment (thompsonmj) [inline example_scripts/submit_mpi_download.sh:1]: Need `source "${REPO_ROOT}/config/hpc.env"` under `REPO_ROOT=` now that we're specifying `DOWNLOADER_MAX_NODES`, etc. in there.

File: example_scripts/submit_mpi_download.sh [hunk 2]
```diff
@@ -23,15 +23,15 @@ if [ "$4" != "" ] && [ "$4" != "--recheck" ]; then
     base_filename=$(basename "${filename}" ."${ext}")
 
     # Submit the script to Slurm
-    sbatch --output="${LOGS_DIR}/${base_filename}.out" --error="${LOGS_DIR}/${base_filename}.err" --dependency=afterany:"${dependency}" "${script}" "${schedule}" "${iteration_number}" "$5"
+    sbatch --output="${LOGS_DIR}/${base_filename}.out" --error="${LOGS_DIR}/${base_filename}.err" --dependency=afterany:"${dependency}" "${script}" "${schedule}" "${iteration_number}" "$5" --nodes="${DOWNLOADER_MAX_NODES}" --ntasks-per-node="${DOWNLOADER_WORKERS_PER_NODE}" --cpus-per-task="${DOWNLOADER_CPU_PER_TASK}" --account="${ACCOUNT}"
```
Comment (thompsonmj) [inline example_scripts/submit_mpi_download.sh:26]:
```suggestion
    sbatch \
        --output="${LOGS_DIR}/${base_filename}.out" \
        --error="${LOGS_DIR}/${base_filename}.err" \
        --dependency=afterany:"${dependency}" \
        "${script}" "${schedule}" "${iteration_number}" "$5" \
        --nodes="${DOWNLOADER_MAX_NODES}" \
        --ntasks-per-node="${DOWNLOADER_WORKERS_PER_NODE}" \
        --cpus-per-task="${DOWNLOADER_CPU_PER_TASK}" \
        --account="${ACCOUNT}"
```

Comment (thompsonmj) [reply example_scripts/submit_mpi_download.sh:26]: To help readability.

File: example_scripts/submit_mpi_download.sh [hunk 3]
```diff
@@ -23,15 +23,15 @@ if [ "$4" != "" ] && [ "$4" != "--recheck" ]; then
...
     exit 0
 else
     filename=$(basename "$script")
     ext="${filename##*.}"
     base_filename=$(basename "${filename}" ."${ext}")
 
     # Submit the script to Slurm
-    sbatch --output="${LOGS_DIR}/${base_filename}.out" --error="${LOGS_DIR}/${base_filename}.err" "${script}" "${schedule}" "${iteration_number}" "$4"
+    sbatch --output="${LOGS_DIR}/${base_filename}.out" --error="${LOGS_DIR}/${base_filename}.err" "${script}" "${schedule}" "${iteration_number}" "$4" --nodes="${DOWNLOADER_MAX_NODES}" --ntasks-per-node="${DOWNLOADER_WORKERS_PER_NODE}" --cpus-per-task="${DOWNLOADER_CPU_PER_TASK}" --account="${ACCOUNT}"
```
Comment (thompsonmj) [inline example_scripts/submit_mpi_download.sh:34]:
```suggestion
    sbatch \
        --output="${LOGS_DIR}/${base_filename}.out" \
        --error="${LOGS_DIR}/${base_filename}.err" \
        "${script}" "${schedule}" "${iteration_number}" "$4" \
        --nodes="${DOWNLOADER_MAX_NODES}" \
        --ntasks-per-node="${DOWNLOADER_WORKERS_PER_NODE}" \
        --cpus-per-task="${DOWNLOADER_CPU_PER_TASK}" \
        --account="${ACCOUNT}"
```

**Review by thompsonmj — CHANGES_REQUESTED** [review]
Only suggestion affecting function is to source the dotenv file in the bash script.

**Comment by egrace479** [comment]
There are still some questions to be answered on overall package and things to do after this update is settled:
 1. Presumably we will want to change `mpi_downloader` to `distributed_downloader` in `src`. We then have to decide how to work with the other scripts outside of it but still in `src`. The idea is that `src/submitter.py` will be the interface that runs the whole download, while the `scripts/` and `config/` files will be modified by users to pass into the process.
 2. The resize functions should be moved out to a `tools` folder or similar at the root level of the repository. I'm fine with leaving them not generalized for now, but good to include the tools.
 3. The functioning of the preprocessing steps should be clarified (i.e., how to run `server_prep` and `MPI_download_prep`--what goes in and what actually comes out). 
 4. Similarly to 3, we need to clarify requirements for the actual download process (if nothing else, this could be the structure established by the scripts noted in 3).
 5. We need tests. In particular, this restructuring should be tested to ensure it still functions as expected.

**Review by Andrey170170 — APPROVED** [review]

File: scripts/server_downloading_prep.slurm [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 source "${REPO_ROOT}/config/osc.env"
```
Comment (thompsonmj) [inline scripts/server_downloading_prep.slurm:6]:
```suggestion
source "${REPO_ROOT}/config/hpc.env"
```

File: scripts/server_profiling.slurm [hunk 5]
```diff
@@ -15,7 +15,7 @@ max_worker_per_node=4
 # ================== DO NOT TOUCH BELOW THIS LINE ==============================
 source "${REPO_ROOT}/config/osc.env"
```
Comment (thompsonmj) [inline scripts/server_profiling.slurm:16]:
```suggestion
source "${REPO_ROOT}/config/hpc.env"
```

**Review by thompsonmj — APPROVED** [review]
2x `osc.env` -> `hpc.env`, but I think it looks alright otherwise. Will be helpful to start getting tests written.

**Pull request merged**